File: Interview Problems/Parking Lot/Java Code/service/ReceiptService.java
package service;

import domain.Receipt;
import domain.Ticket;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

public class ReceiptService {

    private static final DateTimeFormatter FORMATTER =
            DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss");

    public ReceiptService() {
        System.out.println("[SERVICE] ReceiptService initialized");
    }

    public Receipt generateReceipt(Ticket ticket, double fee) {
        System.out.println("[SERVICE] Generating receipt for ticket: " + ticket.getId());
        Receipt receipt = new Receipt(ticket.getId(), fee);
        System.out.println("[SERVICE] Receipt generated: " + receipt.getId() + " with fee: ₹" + fee);
        return receipt;
    }

    public void markReceiptAsPaid(Receipt receipt) {
        System.out.println("[SERVICE] Marking receipt as paid: " + receipt.getId());
        receipt.markAsPaid();
        System.out.println("[SERVICE] Receipt marked as paid successfully");
    }

    public String generateReceiptText(Receipt receipt, Ticket ticket) {
        System.out.println("[SERVICE] Generating receipt text for: " + receipt.getId());

        // Use the receipt's own exitTime (set at construction) for consistency
        LocalDateTime exitTime = receipt.getExitTime();
        long minutesParked = ChronoUnit.MINUTES.between(ticket.getEntryTime(), exitTime);
        long hoursParked = minutesParked / 60;
        long remainingMinutes = minutesParked % 60;

        StringBuilder receiptText = new StringBuilder();
        receiptText.append("\n");
        receiptText.append("=====================================\n");
        receiptText.append("         \uD83E\uDDFE PARKING RECEIPT         \n");
        receiptText.append("=====================================\n");
        receiptText.append(String.format("%-14s: %s%n", "Receipt ID",  receipt.getId()));
        receiptText.append(String.format("%-14s: %s%n", "Ticket ID",   ticket.getId()));
        receiptText.append(String.format("%-14s: %s%n", "Slot ID",     ticket.getSlotId()));
        receiptText.append("-------------------------------------\n");
        receiptText.append(String.format("%-14s: %s%n", "Entry Time",  ticket.getEntryTime().format(FORMATTER)));
        receiptText.append(String.format("%-14s: %s%n", "Exit Time",   exitTime.format(FORMATTER)));
        receiptText.append(String.format("%-14s: %dh %dm%n", "Duration", hoursParked, remainingMinutes));
        receiptText.append("-------------------------------------\n");
        receiptText.append(String.format("%-14s: \u20B9%.2f%n", "Total Fee",     receipt.getTotalFee()));
        receiptText.append(String.format("%-14s: %s%n",  "Status",         receipt.getPaymentStatus()));
        receiptText.append("=====================================\n");
        receiptText.append("   Thank you for parking with us!   \n");
        receiptText.append("=====================================\n");

        System.out.println("[SERVICE] Receipt text generated successfully");
        return receiptText.toString();
    }
}

File: Interview Problems/Parking Lot/Python Code/service/receipt_service.py
import uuid
from datetime import datetime
from domain.ticket import Ticket
from domain.receipt import Receipt

class ReceiptService:
    DATE_FORMAT = "%d-%m-%Y %H:%M:%S"

    def __init__(self):
        print("[SERVICE] ReceiptService initialized")

    def generate_receipt(self, ticket: Ticket, fee: float) -> Receipt:
        print(f"[SERVICE] Generating receipt for ticket: {ticket.id}")
        receipt = Receipt(ticket.id, fee)
        print(f"[SERVICE] Receipt generated: {receipt.id} with fee: ₹{fee}")
        return receipt

    def mark_receipt_as_paid(self, receipt: Receipt):
        print(f"[SERVICE] Marking receipt as paid: {receipt.id}")
        receipt.mark_as_paid()
        print("[SERVICE] Receipt marked as paid successfully")

    def generate_receipt_text(self, receipt: Receipt, ticket: Ticket) -> str:
        print(f"[SERVICE] Generating receipt text for: {receipt.id}")
        exit_time = receipt.exit_time
        duration = exit_time - ticket.entry_time
        minutes_parked = int(duration.total_seconds() / 60)
        hours_parked = minutes_parked // 60
        remaining_minutes = minutes_parked % 60

        lines = [
            "",
            "=====================================",
            "         🧾 PARKING RECEIPT         ",
            "=====================================",
            f"{'Receipt ID':<14}: {receipt.id}",
            f"{'Ticket ID':<14}: {ticket.id}",
            f"{'Slot ID':<14}: {ticket.slot_id}",
            "-------------------------------------",
            f"{'Entry Time':<14}: {ticket.entry_time.strftime(self.DATE_FORMAT)}",
            f"{'Exit Time':<14}: {exit_time.strftime(self.DATE_FORMAT)}",
            f"{'Duration':<14}: {hours_parked}h {remaining_minutes}m",
            "-------------------------------------",
            f"{'Total Fee':<14}: ₹{receipt.total_fee:.2f}",
            f"{'Status':<14}: {receipt.payment_status.name}",
            "=====================================",
            "   Thank you for parking with us!   ",
            "=====================================",
            ""
        ]
        print("[SERVICE] Receipt text generated successfully")
        return "\n".join(lines)
